.github/workflows/pipeline.yml

```diff
@@ -0,0 +1,40 @@
+name: Tests
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  # build with different go versions
+  build:
+    name: "\U0001F528 Build / :go: ${{ matrix.go }}"
+    strategy:
+      fail-fast: false
+      matrix:
+        go: ["1.11", "1.12"]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-go@v3
+        with:
+          go-version: ${{ matrix.go }}
+      - run: go build
+
+  lint:
+    name: "\U0001F9F9 Run tests and linting"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-go@v3
+        with:
+          go-version: 1.12
+      - run: |-
+          go get golang.org/x/lint/golint
+          go get github.com/fzipp/gocyclo@v0.0.0-20150627053110-6acd4345c835
```

Comment:
seems like the current version of gocyclo is unhappy with the old version of go here.

Maybe I should just update the go version in use?

Reply:
It may be easier to use `golangci-lint`, i.e. by copying the `.golangci.yml` file from the Dendrite repo and then doing something like this:
```
  lint:
    timeout-minutes: 5
    name: Linting
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Go
        uses: actions/setup-go@v3
        with:
          go-version: 1.18
      - name: golangci-lint
        uses: golangci/golangci-lint-action@v3
```

We're Go 1.18 minimum across Dendrite, gomatrixserverlib, Pinecone etc, so it might make sense to do the same here too.

The federation tester is minimum 1.15 in `go.mod` but the Docker image is actually built with Go 1.19 so it should be safe to increase that too.

Reply:
and suddenly this becomes a bigger change than I wanted to do today. Changing to golangci-lint means figuring out which of the things currently done by lint.sh it does.